### PR TITLE
utils/network: Big refactoring on network library to improve consistency [v2]

### PR DESCRIPTION
**Pull request by beraldoleal**
Multiple pull requests were trying to add more features to this library.
But because of a lack of convention, it was getting hard to understand
this and we were losing usability. Besides a complete rewrite here, this
pull request is a) adding a complete set of methods to handle network
interfaces parameters; b) adding a wait when setting MTU values on
interfaces; and c) better support for IPv6.

Thanks for @anascko, @vaishnavibhat, @PraveenPenguin, and @smruti77

-----
#### Changes from v1:

  * Fixed _connect() method to raise exceptions when fails

### CODE REVIEW

File: avocado/utils/network/hosts.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+# Copyright: 2019-2020 Red Hat Inc.
+# Authors : Beraldo Leal <bleal@redhat.com>
+
+"""
+This module provides an useful API for hosts in a network.
+"""
+
+from ..ssh import Session
+
+from .common import _run_command
```
Comment (apahim) [inline avocado/utils/network/hosts.py]: should this still be a private function?

File: avocado/utils/network/hosts.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    A host can be local or remote. If you pass port, username, key or password,
+    a connection will attempt to be created.
+
+    During the initialization, all interfaces will be detected and available
+    via `interfaces` attribute.
+
+    So, for instance you could have a local and a remote host::
+
+        local = Host(host='foo', port=22, username='foo', password='bar')
+        remote = Host(host='bar')
```
Comment (apahim) [inline avocado/utils/network/hosts.py]: local and remote are inverted here. Providing a username is what enables the remote feature.

File: avocado/utils/network/hosts.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+from .common import _run_command
+from .interfaces import NetworkInterface
+from .exceptions import NWException
+
+
+class Host:
+    """
+    This class represents a Host.
+
+    A host can be local or remote. If you pass port, username, key or password,
```
Comment (apahim) [inline avocado/utils/network/hosts.py]: Only username is enough to create the connection.

File: avocado/utils/network/hosts.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        self.remote_session = self._connect()
+
+    def _connect(self):
+        if self.username:
+            session = Session(host=self.host,
+                              port=self.port,
+                              user=self.username,
+                              key=self.key,
+                              password=self.password)
+            if session.connect():
```
Comment (apahim) [inline avocado/utils/network/hosts.py]: No sure this is a good idea.... I'd imagine that users would create the instance in the test `setUp()` and only use it durint the test methods. But we can start like this and improve later, if that becomes a problem.

File: avocado/utils/network/interfaces.py [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+    Here you will find a few methods to perform basic operations on a NIC.
+    """
+
+    def __init__(self, if_name, if_type='Ethernet', remote_session=None):
+        self.name = if_name
+        self.if_type = if_type
+        self.remote_session = remote_session
+
+    def _get_interface_details(self, version=4):
+        output = ''
```
Comment (apahim) [inline avocado/utils/network/interfaces.py]: leftover?

**Comment by beraldoleal** [comment]
> Looks very nice, thank you for the port fix.
> The username handling seems still a bit inconsistent:
> 
> ```
> >>> h = Host(host='192.168.0.94', username='root', port=22, password=password)
> >>> [iface.name for iface in h.interfaces]
> ['eth0', 'lo', 'wlan0']
> 
> >>> h = Host(host='192.168.0.94', username='foobar', port=22, password=password)
> Traceback (most recent call last):
>   File "", line 1, in 
>   File "/home/apahim/git/avocado/avocado/utils/network/hosts.py", line 55, in __init__
>     self.remote_session = self._connect()
>   File "/home/apahim/git/avocado/avocado/utils/network/hosts.py", line 68, in _connect
>     raise NWException(msg)
> avocado.utils.network.exceptions.NWException: Failed connecting 192.168.0.94:22
> 
> >>> h = Host(host='192.168.0.94', username=' ', port=22, password=password)
> Traceback (most recent call last):
>   File "", line 1, in 
>   File "/home/apahim/git/avocado/avocado/utils/network/hosts.py", line 55, in __init__
>     self.remote_session = self._connect()
>   File "/home/apahim/git/avocado/avocado/utils/network/hosts.py", line 68, in _connect
>     raise NWException(msg)
> avocado.utils.network.exceptions.NWException: Failed connecting 192.168.0.94:22
> 
> >>> h = Host(host='192.168.0.94', username='', port=22, password=password)
> >>> [iface.name for iface in h.interfaces]
> ['br-7c16cfa0b398', 'docker0', 'enp0s31f6', 'ens1u1', 'lo', 'tun0', 'virbr0', 'virbr0-nic', 'wlp0s20f3']
> ```
> 
> but I'm ok with fixing that later.

Yes, About this

File: avocado/utils/network/interfaces.py (hunk 5 above)
Comment (beraldoleal) [reply avocado/utils/network/interfaces.py]: Yes, fixed. Good catch.

File: avocado/utils/network/hosts.py (hunk 2 above)
Comment (beraldoleal) [reply avocado/utils/network/hosts.py]: Thanks. Fixed.

File: avocado/utils/network/hosts.py (hunk 1 above)
Comment (beraldoleal) [reply avocado/utils/network/hosts.py]: You are right, fixed.

**Comment by beraldoleal** [comment]
@apahim thanks for your comments. I just implemented a new way to better describe a Remote vs a Local host using inheritance.

Also, aplied your comments.

Closing this in favor of a v2.

**Pull request closed without merging**